[Accessibility] Update show password icon to use a button element

darrenjacoby opened the pull request:
### Submission Review Guidelines:

- * I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- * I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- * I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- * Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #51834  .
Closes #51826 .
Closes #51827 .
Closes #51825 .
Closes #51835 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Account:**

1. Browse to My Account → Account Details.
2. The Password Change form should display the Show/Hide Password icon as a `<button>` element.
3. The user should be able to tab to the Show/Hide Password button.
4. The button should show a focus state.
5. The button should have an `aria-label` that describes the action for screen readers.

**Login:**

1. Browse to My Account → Login.
2. The Password Change form should display the Show/Hide Password icon as a `<button>` element.
3. The user should be able to tab to the Show/Hide Password button.
4. The button should show a focus state.
5. The button should have an `aria-label` that describes the action for screen readers.

**Other dev notes:**

- I've updated the focus state once the Show/Hide Password button has been actioned to focus back to the input field. Let me know if this is considered favourable.
- I've added `aria-describedBy` for the Show/Hide Password button to link back to the relevant input field.
- I've updated themes Twenty Seventeen, Twenty Twenty and Twenty Twenty-One to avoid style regressions.
- Noting that Storefront could use `background-color: transparent` on the `.show-password-input` class. Do we have recommendation of where we would want to add that?

### Changelog entry

- * [ ] Automatically create a changelog entry from the details below.
- * [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance
#### Type

</details>

darrenjacoby committed:
Update show password input to use a button element

darrenjacoby committed:
Add backward compatibility styles

darrenjacoby committed:
Add changelog

darrenjacoby commented:
@dkotter ready for your review, left a few dev notes above, let me know your thoughts.

dkotter commented on plugins/woocommerce/client/legacy/js/frontend/woocommerce.js:
```diff
@@ -104,26 +104,36 @@ jQuery( function ( $ ) {
+		
+			if ( $( this ).hasClass( 'display-password' ) ) {
+				$( this ).removeClass( 'display-password' );
+				$( this ).attr( 'aria-label', 'Show password' );
+			} else {
+				$( this ).addClass( 'display-password' );
+				$( this ).attr( 'aria-label', 'Hide password' );
```
I believe we'll want these `aria-labels` to be translated, whereas right now they are hardcoded to the English version

dkotter commented on plugins/woocommerce/client/legacy/js/frontend/woocommerce.js:
```diff
@@ -104,26 +104,36 @@ jQuery( function ( $ ) {
+		$('.password-input').each(function() {
+			const describedBy = $(this).find('input').attr('id');
+			$(this).append(
+				'<button class="show-password-input" aria-label="Show password" aria-describedBy="' + describedBy + '"></button>'
```
Same comment on this label, it will need to be passed through translation

darrenjacoby committed:
Add i18n password show and hide

darrenjacoby commented:
@dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.

dkotter commented:
> @dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.

That works but also means that data will be added on all page views I believe. Ideally we only add this extra information on pages that need it (which seems like the My Account pages only). I'm not sure if there's a script that only loads on those pages that we could attach to? I see in the `get_script_data` method there's a `wc-password-strength-meter` script, could adding these values there work?

darrenjacoby committed:
Move i18n params to wc_password_strength_meter

darrenjacoby commented:
> > @dkotter good point, I've updated the PR to add `woocommerce.i18n_password_show` and `woocommerce.i18n_password_hide`. I'm not sure if this is a desirable structure (under namespace woocommerce), so let me know your thoughts.
> 
> That works but also means that data will be added on all page views I believe. Ideally we only add this extra information on pages that need it (which seems like the My Account pages only). I'm not sure if there's a script that only loads on those pages that we could attach to? I see in the `get_script_data` method there's a `wc-password-strength-meter` script, could adding these values there work?

Yea, that works too, and as you mentioned would fire only on the required pages. I've updated that code. The semantics aren't "exact", but we can rename `wc_password_strength_meter_params` if it's an issue.

dkotter reviewed (approved):
@ankitguptaindia @qasumitbagthariya This is ready for QA

darrenjacoby committed:
Revert i18n params

darrenjacoby commented:
Hi @ankitguptaindia , thanks for catching that. @dkotter it's a regression from moving the params, I've moved them back to `woocommerce` which feels correct as the referenced code lives within `woocommerce.js` as well.

ankitguptaindia commented:
**QA/Test Report-** ✅ 
------------

**Testing Environment -** 

<Details>

* WordPress: 6.7.1
* Theme:  
  - Storefront Version: 4.6.0  
  - Twenty Twenty-Four Version: 1.3  
  - Twenty Twenty-Five Version: 1.0
* WooCommerce - Development version with this fix-specific branch.  
* PHP: 8.3
* Web Server: Nginx
* Browser: Chrome - Version 131
* OS: macOS Sonoma 14.5


</Details>


**Test Results -** 

The changes work as intended and tested across Safari, Firefox, and Chrome. Accessibility was verified using macOS VoiceOver Version 10. Testing was conducted with the Twenty Twenty-Five, Storefront theme and Twenty Twenty-Four theme.
 

**Functional Demo /  Screencast -**

**With macOS Voice Over:**

https://github.com/user-attachments/assets/3d01a602-d3b6-4355-9050-a3470e6b1ae5

ankitguptaindia commented:
Thanks for the great work on this PR @darrenjacoby would you mind updating this PR status to open for Woo CR? Thank you! 

cc: @vikrampm1

vikrampm1 commented:
@mikejolley this PR is ready for your review!

mikejolley reviewed (approved):
This looks good to me. Tested across a few default themes. 

For storefront I think the fix to the color would need to be in that theme because it dequeues core stylesheets, unless we want to inline some basic styles.

darrenjacoby commented:
@mikejolley yea, we should handle any storefront style changes as a separate PR on the storefront repo. It does however render quite well without any changes;

<img width="621" alt="image" src="https://github.com/user-attachments/assets/ba2d91f7-aad4-4dcd-a16e-2ccab5aeb1be">

vikrampm1 commented:
@mikejolley if there are no further questions then please proceed and merge the PR, thanks!

mikejolley commented:
There were some linting errors after running CI (https://github.com/woocommerce/woocommerce/pull/53489). Lints don't run on forks. I can take care of them assuming the rest of the CI passes.

vikrampm1 commented:
@darrenjacoby this requires further updates before it can be merged, so please make the necessary changes and then notify Mike to merge the PR.

darrenjacoby commented:
@mikejolley that is a legitimate flag, and something I noticed as well when testing. While this improves accessibility, it may impact user experience negatively. The user behaviour, based on the elements, is expected. It may require more discussions with the accessibility team and their recommendations in such cases. cc @vikrampm1

amberhinds commented:
RE @nielslange's comment
> Before this PR, a user could navigate to My Account → Login, tab into the username field, press Enter, provide their password, and press Enter again to log in. With this PR, pressing Enter after providing the password toggles the password visibility instead. As a result, pressing Enter no longer submits the form.

I agree that the loss of Return key functionality in the password field to submit the form may frustrate users who are used to it working this way. I'm not certain that it's a problem to remove that functionality, though - it's more of an annoyance. I asked our accessibility specialist for her opinion as well.

Is it possible to add it back and still address the password visibility button?

amberhinds commented:
We discussed this, and removing the ability to submit the form by pressing return in the password field does not create an accessibility issue, so if it's not possible to add that functionality back, I don't think it's a problem. It's better to have a keyboard-accessibility button for password visibility.

darrenjacoby committed:
Fix linting issues

darrenjacoby committed:
Handle enter key

darrenjacoby commented:
@amberhinds @mikejolley @nielslange let me know if the implementation above could work.

It disables the "Enter" key for the password field (I don't think we should show the password if the user presses "Enter", it feels like it could pose an security risk). For the login form, pressing Enter on the password field triggers a pseudo event of clicking "Submit".

Please let me know your thoughts if this feels viable.

nielslange reviewed (approved):
I’ve tested all the changes successfully, @darrenjacoby. Thanks again for your work on this! I'll merge the PR once all CI checks have passed.

Regarding https://github.com/woocommerce/woocommerce/pull/53098#issuecomment-2519754633, could you create an issue in https://github.com/woocommerce/storefront/issues?

nielslange committed:
Merge branch 'trunk' into fix/51834

nielslange commented on plugins/woocommerce/includes/class-wc-frontend-scripts.php:
```diff
@@ -497,6 +497,8 @@ private static function get_script_data( $handle ) {
 				$params = array(
 					'ajax_url'    => WC()->ajax_url(),
 					'wc_ajax_url' => WC_AJAX::get_endpoint( '%%endpoint%%' ),
+					'i18n_password_show'    => esc_attr__( 'Show password', 'woocommerce' ),
+					'i18n_password_hide'    => esc_attr__( 'Hide password', 'woocommerce' ),
 				);
```
```suggestion
				$params = array(
					'ajax_url'           => WC()->ajax_url(),
					'wc_ajax_url'        => WC_AJAX::get_endpoint( '%%endpoint%%' ),
					'i18n_password_show' => esc_attr__( 'Show password', 'woocommerce' ),
					'i18n_password_hide' => esc_attr__( 'Hide password', 'woocommerce' ),
				);
```

nielslange replied:
While running the CI checks, I noticed the following lint error.

Commit:
Update plugins/woocommerce/includes/class-wc-frontend-scripts.php

amberhinds commented:
Sorry, I'm late to the game. This seems like a good solution to me. I agree that unexpectedly showing someone's password could be a security issue.
